File: server/src/services/objects/Authorizator.cpp
#include "Authorizator.hpp"
#include "RegisteredUser.hpp"
#include "services/db/UserService.hpp"

nscalc::UserData AuthorizatorImpl::tryLogIn(std::string_view user_email, std::string_view user_password) {
  std::lock_guard lk(*userService_);
  if (auto user = userService_->getUserByEmail(user_email); user) {
    if (!user_password.empty() && userService_->checkPassword(*user, user_password) == false) {
      dataObservers_->alarm(nscalc::AlarmType::Warning, "User \"" + user->user_name + "\" forgot his password");
      throw nscalc::AuthorizationFailed(nscalc::AuthorizationFailed_Reason::incorrect_password);
    }

    auto new_user = new RegisteredUser(*user, solutionService_, fertilizerService_, calculationService_, dataObservers_);
    auto oid = user_poa->activate_object(new_user, 
      nprpc::ObjectActivationFlags::SESSION_SPECIFIC,
      &nprpc::get_context()
    );

    Session s;
    do {
      s.sid = UserService::create_uuid();
    } while (sessions_.find(s) != sessions_.end());
    s.email = user_email;

    dataObservers_->alarm(nscalc::AlarmType::Info, "User \"" + user->user_name + "\" has logged in");

    return {user->user_name, sessions_.emplace(std::move(s)).first->sid, oid};
  } else {
    throw nscalc::AuthorizationFailed(nscalc::AuthorizationFailed_Reason::email_does_not_exist);
  }
}

// Implementation of the IAuthorizator_Servant interface 
nscalc::UserData AuthorizatorImpl::LogIn(::nprpc::flat::Span<char> login, ::nprpc::flat::Span<char> password) {
  auto user_email = (std::string_view)login;
  auto user_password = (std::string_view)password;

  if (user_password.empty()) {
    dataObservers_->alarm(nscalc::AlarmType::Warning, "User is trying to log in with an empty password");
    throw nscalc::AuthorizationFailed(nscalc::AuthorizationFailed_Reason::incorrect_password);
  }

  return tryLogIn(user_email, user_password);
}

// Implementation of the IAuthorizator_Servant interface 
nscalc::UserData AuthorizatorImpl::LogInWithSessionId(::nprpc::flat::Span<char> session_id) {
  auto sid = (std::string_view)session_id;
  Session s;
  s.sid.assign(sid.data(), sid.size());

  std::string email;
  {
    std::lock_guard lk(sessionsMutex_);
    if (auto it = sessions_.find(s); it != sessions_.end()) {
      email = it->email;
    }
  }

  if (email.empty())
    throw nscalc::AuthorizationFailed(nscalc::AuthorizationFailed_Reason::session_does_not_exist);

  return tryLogIn(email, {});
}

// Implementation of the IAuthorizator_Servant interface 
bool AuthorizatorImpl::LogOut(::nprpc::flat::Span<char> session_id) {
  auto sid = (std::string_view)session_id;
  Session s;
  s.sid.assign(sid.data(), sid.size());

  std::lock_guard<std::mutex> lk(sessionsMutex_);
  return static_cast<bool>(sessions_.erase(s));
}

// Implementation of the IAuthorizator_Servant interface 
bool AuthorizatorImpl::CheckUsername(::nprpc::flat::Span<char> username) {
  std::lock_guard lk(*userService_);
  return userService_->checkUsername(username);
}

// Implementation of the IAuthorizator_Servant interface 
bool AuthorizatorImpl::CheckEmail(::nprpc::flat::Span<char> email) {
  std::lock_guard lk(*userService_);
  return userService_->checkEmail(email);
}

// Implementation of the IAuthorizator_Servant interface 
void AuthorizatorImpl::AuthorizatorImpl::RegisterStepOne(::nprpc::flat::Span<char> username, ::nprpc::flat::Span<char> email, ::nprpc::flat::Span<char> password) {
  {
    std::lock_guard lk(*userService_);
    if (userService_->checkUsername(username) == false)
      throw nscalc::RegistrationFailed(nscalc::RegistrationFailed_Reason::username_already_exist);
    if (userService_->checkEmail(email) == false)
      throw nscalc::RegistrationFailed(nscalc::RegistrationFailed_Reason::email_already_registered);
  }

  NewUser user;
  user.user = std::make_unique<UserService::User>();

  user.user->password_sha256 = UserService::sha256(password);
  user.user->user_name = username;
  user.user->email = email;

  std::uniform_int_distribution<std::uint32_t> dist(10000, 99999);
  user.code = dist(rd_);

  // std::cerr << "Code: " << user.code << '\n';

  dataObservers_->alarm(nscalc::AlarmType::Info, "Dear " + user.user->user_name + ", here is your confirmation code: " + std::to_string(user.code));

  {
    std::lock_guard<std::mutex> lk(new_users_mut_);
    new_users_db_.emplace(user.user->user_name, std::move(user));
  }
}

// Implementation of the IAuthorizator_Servant interface 
void AuthorizatorImpl::RegisterStepTwo(::nprpc::flat::Span<char> username, uint32_t code) {
  std::lock_guard<std::mutex> lk(new_users_mut_);
  if (auto it = new_users_db_.find((std::string)username); it != new_users_db_.end()) {
    if (it->second.code != code) {
      throw nscalc::RegistrationFailed(nscalc::RegistrationFailed_Reason::incorrect_code);
    }
    dataObservers_->alarm(nscalc::AlarmType::Info, "New user '" + it->first + "' has been registered");
    auto user = std::move(it->second.user);
    new_users_db_.erase(it);
    {
      std::lock_guard lk(*userService_);
      userService_->addUser(std::move(user));
    }
  } else {
    throw nscalc::RegistrationFailed(nscalc::RegistrationFailed_Reason::invalid_username);
  }
}

AuthorizatorImpl::AuthorizatorImpl(
  nprpc::Rpc& rpc,
  std::shared_ptr<UserService> userService,
  std::shared_ptr<SolutionService> solutionService,
  std::shared_ptr<FertilizerService> fertilizerService,
  std::shared_ptr<CalculationService> calculationService,
  std::shared_ptr<DataObservers> dataObservers)
  : userService_{userService}
  , solutionService_{solutionService}
  , fertilizerService_{fertilizerService}
  , calculationService_{calculationService}
  , dataObservers_{dataObservers}
{	
  user_poa = nprpc::PoaBuilder(&rpc)
		.with_max_objects(1024)
		.with_lifespan(nprpc::PoaPolicy::Lifespan::Transient)
		.build();

  userService_->load();
}

File: server/Sources/NScalcServer/Authorizator.swift
import Foundation
import NPRPC
import NScalc

final class AuthorizatorImpl: AuthorizatorServant, @unchecked Sendable {
	private struct PendingRegistration: Sendable {
		let username: String
		let email: String
		let password: String
		let code: UInt32
	}

	private let userPoa: Poa
	private let userService: UserService
	private let solutionService: SolutionService
	private let fertilizerService: FertilizerService
	private let calculationService: CalculationService
	private let sessionService: SessionService
	private let pendingLock = NSLock()
	private var pendingRegistrations: [String: PendingRegistration] = [:]

	init(rpc: Rpc, db: AppDatabase) throws {
		self.userPoa = try rpc.createPoa(maxObjects: 1024, lifetime: .transient, idPolicy: .systemGenerated)
		self.userService = UserService(db: db)
		self.solutionService = SolutionService(db: db)
		self.fertilizerService = FertilizerService(db: db)
		self.calculationService = CalculationService(db: db)
		self.sessionService = SessionService(db: db)
		super.init()
	}

	private func makeUserData(for user: UserRecord) throws -> UserData {
		guard let userId = user.id else {
			throw RuntimeError(message: "User record has no id")
		}

		let servant = RegisteredUserImpl(
			user: user,
			solutionService: solutionService,
			fertilizerService: fertilizerService,
			calculationService: calculationService
		)

		var flags: ObjectActivationFlags = .allowAll
		if sessionContext != nil {
			flags.insert(.privateSession)
		}

		let oid = try userPoa.activateObject(servant, flags: flags, sessionContext: sessionContext)
		guard let userObject = NPRPCObject.fromObjectId(oid) else {
			throw RuntimeError(message: "Failed to create RegisteredUser object")
		}

		let sessionId = try sessionService.createSession(userId: userId)
		return UserData(name: user.name, isAdmin: user.isAdmin, sessionId: sessionId, db: userObject)
	}

	private func loginByEmail(_ email: String, password: String?) throws -> UserData {
		let normalizedEmail = email.lowercased()
		guard let user = try userService.user(byEmail: normalizedEmail) else {
			throw AuthorizationFailed(reason: .email_does_not_exist)
		}

		if let password, !password.isEmpty, !user.verifyPassword(password) {
			throw AuthorizationFailed(reason: .incorrect_password)
		}

		return try makeUserData(for: user)
	}

	override func logIn(login: String, password: String) throws -> UserData {
		guard !password.isEmpty else {
			throw AuthorizationFailed(reason: .incorrect_password)
		}
		return try loginByEmail(login, password: password)
	}

	override func logInWithSessionId(session_id: String) throws -> UserData {
		guard let user = try sessionService.user(forSession: session_id) else {
			throw AuthorizationFailed(reason: .session_does_not_exist)
		}
		return try makeUserData(for: user)
	}

	override func logOut(session_id: String) -> Bool {
		do {
			guard try sessionService.user(forSession: session_id) != nil else {
				return false
			}
			try sessionService.deleteSession(session_id)
			return true
		} catch {
			print("[Authorizator] logOut failed: \(error)")
			return false
		}
	}

	override func checkUsername(username: String) -> Bool {
		do {
			return try userService.isUsernameAvailable(username)
		} catch {
			print("[Authorizator] checkUsername failed: \(error)")
			return false
		}
	}

	override func checkEmail(email: String) -> Bool {
		do {
			return try userService.isEmailAvailable(email)
		} catch {
			print("[Authorizator] checkEmail failed: \(error)")
			return false
		}
	}

	override func registerStepOne(username: String, email: String, password: String) throws {
		if !(try userService.isUsernameAvailable(username)) {
			throw RegistrationFailed(reason: .username_already_exist)
		}
		if !(try userService.isEmailAvailable(email)) {
			throw RegistrationFailed(reason: .email_already_registered)
		}

		let code = UInt32.random(in: 10000...99999)
		let pending = PendingRegistration(
			username: username,
			email: email.lowercased(),
			password: password,
			code: code
		)

		pendingLock.lock()
		pendingRegistrations[username] = pending
		pendingLock.unlock()

		print("[Authorizator] registration code for \(username): \(code)")
	}

	override func registerStepTwo(username: String, code: UInt32) throws {
		let pending: PendingRegistration?
		pendingLock.lock()
		pending = pendingRegistrations[username]
		pendingLock.unlock()

		guard let pending else {
			throw RegistrationFailed(reason: .invalid_username)
		}
		guard pending.code == code else {
			throw RegistrationFailed(reason: .incorrect_code)
		}

		_ = try userService.addUser(name: pending.username, email: pending.email, password: pending.password)

		pendingLock.lock()
		pendingRegistrations.removeValue(forKey: username)
		pendingLock.unlock()
	}
}